creator/middleware.py

```diff
@@ -104,3 +111,96 @@ def _get_new_key():
+    """
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        # Only attempt to resolve if user has not yet been extracted
+        if not request.user.is_authenticated:
+            request.user = self.__class__.get_jwt_user(request)
+        return self.get_response(request)
+
+    @staticmethod
+    def get_jwt_user(request):
+        """
+        Creates a user object from the JWT in the Authorization header.
+        This user object will be used to grant permissions based on their
+        groups and roles.
+        If no valid token is found, an anonymous user will be returned
+
+        If running with the DEVELOP = True setting, all requests will be
+        treated as requests from an admin to grant permission to all data
+        """
+        if settings.DEVELOP:
+            # Assume user is admin if running in dev mode
+            return get_user_model()(
+                    ego_groups=[],
+                    ego_roles=['ADMIN'])
+
+        user = request.user
+        if user.is_authenticated:
+            return user
+
+        encoded = request.META.get('HTTP_AUTHORIZATION')
+        if not encoded:
+            return AnonymousUser()
+        encoded = encoded.replace('Bearer ', '')
+
+        try:
+            # Validate JWT using the Auth0 key
+            public_key = Auth0AuthenticationMiddleware._get_auth0_key()
+            token = jwt.decode(encoded, public_key, algorithms='RS256',
+                               audience=settings.AUTH0_AUD)
+        except jwt.exceptions.DecodeError as err:
+            logger.error(f'Problem authenticating request: {err}')
+            return AnonymousUser()
+        except jwt.exceptions.InvalidTokenError as err:
+            logger.error(f'Token provided is not valid: {err}')
+            return AnonymousUser()
+
+        groups = token.get('https://kidsfirstdrc.org/groups')
+        roles = token.get('https://kidsfirstdrc.org/roles')
+        # Currently unused
+        permissions = token.get('https://kidsfirstdrc.org/permissions')
+
+        # Don't allow if it looks like the token doesn't have correct fields
+        if groups is None or roles is None:
+            return AnonymousUser()
+
+        user = get_user_model()(
+                ego_groups=groups,
+                ego_roles=roles)
+
+        return user
+
+    @staticmethod
+    def _get_auth0_key():
+        """
+        Attempts to retrieve the auth0 public key from the cache. If it's not
+        there or is expired, fetch a new one from auth0 and store it back in
+        the cache.
+        """
+        key = cache.get(settings.CACHE_AUTH0_KEY, None)
+        # If key is not set in cache (or has timed out), get a new one
+        if key is None:
+            key = Auth0AuthenticationMiddleware._get_new_key()
+            # Cache doesn't like the public key format
+            # cache.set(settings.CACHE_AUTH0_KEY, key,
```

Comment:
The cache doesn't know how to serialize the key returned from `from_jwk()`, will need to look into this further.